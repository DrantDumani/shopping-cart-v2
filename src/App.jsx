import { Outlet } from "react-router";
import Navbar from "./components/Navbar/Navbar";
import { useState } from "react";
import styles from "./App.module.css";
import Icon from "./assets/images/githubIcon.svg?react";

function App() {
  const [cart, setCart] = useState([]);

  const itemAmount = cart.reduce((total, el) => total + el.quantity, 0);

  const addToCart = (obj, num) => {
    const newItem = { ...obj, quantity: num };
    setCart((c) => c.concat(newItem));
  };

  const changeCartQuantity = (id, num = 1) => {
    setCart((c) => {
      let newCart = [...c];
      const itemIndex = newCart.findIndex((item) => item.id === id);
      const targetItem = newCart[itemIndex];
      const newItem = { ...targetItem, quantity: targetItem.quantity + num };
      if (newItem.quantity === 0) newCart = newCart.filter((i) => i.id !== id);
      else newCart[itemIndex] = newItem;
      return newCart;
    });
  };

  const removeItem = (id) => {
    setCart((c) => c.filter((item) => item.id !== id));
  };
  return (
    <>
      <div>
        <header className={styles.headerWrapper}>
          <Navbar amount={itemAmount} />
        </header>
      </div>

      <div className={styles.mainWrapper}>
        <main className={styles.main}>
          <Outlet
            context={{
              addToCart,
              removeItem,
              changeCartQuantity,
              cart,
            }}
          />
        </main>
      </div>

      <div className={styles.footerWrapper}>
        <footer className={styles.footer}>
          <p>
            Created by Darnell.{" "}
            <a href="#">
              Github repo <Icon />
            </a>
          </p>
          <p>
            Built using React Router and{" "}
            <a href="https://fakestoreapi.com/">Fakestore API</a>
          </p>
          <p>Font and homepage image are property of Nintendo.</p>
        </footer>
      </div>
    </>
  );
}

export default App;
