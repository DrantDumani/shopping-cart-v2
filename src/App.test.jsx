/* 
WHAT TO TEST
---
App should render the navbar, footer, and image (just use a snapShot)
Only thing to test should be clicking the navbar and visiting the different routes
Child elements of the root page are just the navbar (the footer doesn't really do anything.)\
BE SURE TO UPDATE THE SNAPSHOT AFTER YOU RUN THIS TEST
*/

import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import App from "./App";

describe("Homepage", () => {
  it("renders navbar, footer, and root route", () => {
    const { container } = render(<App />, { wrapper: MemoryRouter });
    expect(container).toMatchSnapshot();
  });

  it("navigates from the homepage to the shop page", async () => {
    const { container } = render(<App />, { wrapper: MemoryRouter });
    const shopLink = screen.getByRole("link", { name: "Shop" });

    const user = userEvent.setup();
    await user.click(shopLink);

    expect(screen.getAllByRole("heading")[0]).toMatch(/shop/i);
  });
});
