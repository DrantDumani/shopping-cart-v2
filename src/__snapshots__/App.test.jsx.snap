// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Homepage > renders navbar, footer, and root route 1`] = `
<div>
  <header
    class="header"
  >
    <nav
      class="navBar"
    >
      <ul
        class="navList"
      >
        <li>
          <a
            href="/"
          >
            Shellendorf Antiques
          </a>
        </li>
        <li>
          <a
            class=""
            href="/shop"
          >
            Shop
          </a>
        </li>
        <li>
          <a
            class=""
            href="/cart"
          >
            Cart: 
          </a>
        </li>
      </ul>
    </nav>
  </header>
  <main
    class="main"
  />
  <footer
    class="footer"
  >
    <p>
      Created by Darnell. 
      <a
        href="#"
      >
        Github repo
      </a>
    </p>
    <p>
      Font and homepage image are property of Nintendo.
    </p>
    <p>
      Built using React Router and
       
      <a
        href="https://fakestoreapi.com/"
      >
        Fakestore API
      </a>
    </p>
  </footer>
</div>
`;
